Add nanomatch.capture to return captured matches

devongovett opened the pull request:
This adds a new method `nanomatch.capture`, which returns captured matches from dynamic parts of a pattern (e.g. star, globstar). Should partially address https://github.com/micromatch/micromatch/issues/85.

For example:

```javascript
capture('test/**/*.js', 'test/dir/a.js')
// => ['dir', 'a']
```

This is very useful for a number of purposes. My usecase is building a nested object with results of a glob. For example, given the path and pattern above:

```javascript
{dir: {a: 'test/dir/a.js'}}
```

It works by adding a `capture` option to the compiler, which causes capturing groups to be added to the generated regex for stars and globstars. This option is off by default, so everything continues working exactly the same for other methods.

I couldn't get the documentation generator to run on my system, so if you could run it that would be great. Thanks for a great library!

devongovett commented:
Tests don't appear to be passing on windows already... Not sure about that.

If you like this PR, I'll go ahead and add the `capture` method to micromatch as well, and update extglob with support for capture groups as well.

jods4 commented on index.js:
```diff
@@ -535,6 +535,44 @@ nanomatch.matcher = function matcher(pattern, options) {
+        return null;
+      }
+
+      return match.slice(1).filter(Boolean);
```
`.filter(Boolean)` removes the empty captures/strings, right?
Doesn't that make this API hard to use?

For instance: given `a/**/c/**/f.txt` then both paths below result in the same capture `["x"]`:
- `a/x/c/f.txt`
- `a/c/x/f.txt`

It's impossible to rebuild the matching path from the capture.
I you kept empty strings, then the first would capture `["x", ""]` and the second `["", "x"]`. The matching paths are very easy to rebuild.

devongovett replied:
Good point. This made sense for my use case (described above), but I can see it being weird for other cases. I'll remove the filter here and move it to my code.

jonschlinkert commented on index.js:
```diff
@@ -535,6 +535,44 @@ nanomatch.matcher = function matcher(pattern, options) {
+ * @api public
+ */
+
+nanomatch.capture = function(pattern, string, options) {
```
this is a little thing, but could you change `string` to `str` to be consistent with other methods?

jonschlinkert commented on index.js:
```diff
@@ -535,6 +535,44 @@ nanomatch.matcher = function matcher(pattern, options) {
+  var unixify = utils.unixify(options);
+
+  function match() {
+    return function(str) {
```
and change this to `string` (this is probably why you did the other as `string`. noted, and thanks for indulging my OCD)

devongovett commented:
@jonschlinkert updated!

jonschlinkert commented:
@devongovett great work, perfect PR all around! I sent you an invite to join the org. Thank you!

jonschlinkert commented:
I accidentally pushed this up as a patch, since I was focused on a bugfix and I forgot I merged this in first. However, there shouldn't be anything breaking, it's really straight forward, and I tested extensively with micromatch and all tests pass.

devongovett commented:
@jonschlinkert awesome, thanks. I think it should be ok. I tested with micromatch as well, and added all of the nanomatch/extglob tests that I added to micromatch as well: https://github.com/micromatch/micromatch/pull/105.
